bpo-42179: Clarify exception chaining

greatvovan opened the pull request:
Add an important detail and provide a link.

<!-- issue-number: [bpo-42179](https://bugs.python.org/issue42179) -->
https://bugs.python.org/issue42179
<!-- /issue-number -->

greatvovan committed:
Update errors.rst

Clarify exception chaining behaviour and give a reference to the library documentation.

greatvovan committed:
Update errors.rst

Wording

greatvovan committed:
Update errors.rst

Spelling

willingc commented on Doc/tutorial/errors.rst:
```diff
@@ -301,8 +301,13 @@ This can be useful when you are transforming exceptions. For example::
-inside an exception handler or :keyword:`finally` section. Exception chaining
-can be disabled by using ``from None`` idiom:
+inside an exception handler or :keyword:`finally` section, however in this case
+``__context__`` attribute of the exception is used. Chaining with
```
I recommend submitting this sentence right before the builtin exceptions sentence after #23162 lands. Then add "Chaining with ... takes priority."

willingc reviewed (changes requested):
Thanks @greatvovan for the PR. Let's update this after #23162 lands. I like the sentence that I've flagged and think it would be helpful.

greatvovan committed:
Merge branch 'master' into bugfix/bpo-42179

greatvovan commented on Doc/tutorial/errors.rst:
```diff
@@ -300,7 +300,7 @@ This can be useful when you are transforming exceptions. For example::
     RuntimeError: Failed to open database
 
 Exception chaining happens automatically when an exception is raised inside an
-:keyword:`except` or :keyword:`finally` section. Exception chaining can be
```
Tautology.

merwok commented on Doc/tutorial/errors.rst:
```diff
@@ -281,17 +281,17 @@ chaining exceptions. For example::
 
     >>> def func():
-    ...     raise IOError
+    ...     raise ConnectionError
```
Nice!  One OSError was missed on line 294

merwok commented on Doc/tutorial/errors.rst:
```diff
@@ -312,6 +312,13 @@ disabled by using ``from None`` idiom:
+``__context__`` attribute of the new exception with the caught one. When you
+use ``from exc`` syntax, Pyhon fills ``__cause__`` attribute with
+the ``exc``. You can use both attributes for introspection purposes, however
+when formatting exception traceback message, ``__cause__`` takes priority.
```
For the record, I still disagree with this level of implementation notes in the tutorial.

greatvovan replied:
Removed.

greatvovan committed:
Update errors.rst

Remove mentioning of special attributes as folks think it's too much for beginners.

willingc reviewed (approved):
Thanks @greatvovan and good catch on ConnectionError. @merwok If you are good with this as it stands, please merge. Thanks :sunny:
